src/ORM/EagerLoadedList.php

```diff
@@ -0,0 +1,1003 @@
+        // other database implementations could behave differently.
+        $list = $this->foreignID ? $this->dataList->forForeignID($this->foreignID) : $this->dataList;
+        return $list->byIDs($this->column('ID'))->avg($fieldName);
+    }
+
+    /**
+     * Return the sum of the values of the given field in this list
+     *
+     * @param string $fieldName
+     */
+    public function sum($fieldName): int|float
+    {
+        // We have to rely on the database to either give us the right answer or throw the right exception.
+        // MySQL does wiether things with sum, e.g. a sum for the Text field "1_1" is 1 - but
+        // other database implementations could behave differently.
+        $list = $this->foreignID ? $this->dataList->forForeignID($this->foreignID) : $this->dataList;
+        return $list->byIDs($this->column('ID'))->sum($fieldName);
+    }
+
+    /**
+     * Returns true if this list has items
+     */
+    public function exists(): bool
+    {
+        return $this->count() !== 0;
+    }
+
+    public function canFilterBy($fieldName): bool
+    {
+        if (!is_string($fieldName) || empty($this->rows)) {
+            return false;
+        }
+
+        $id = array_key_first($this->rows);
+        return array_key_exists($fieldName, $this->rows[$id]);
+    }
+
+    public function canSortBy($fieldName): bool
+    {
+        return $this->canFilterBy($fieldName);
+    }
+
+    public function find($key, $value): ?DataObject
+    {
+        return $this->filter($key, $value)->first();
+    }
+
+    public function filter(...$args): static
+    {
+        $filters = $this->normaliseFilterArgs($args, __FUNCTION__);
+        $list = clone $this;
+        $list->rows = $this->getMatches($filters);
+        return $list;
+    }
+
+    public function filterAny(...$args): static
+    {
+        $filters = $this->normaliseFilterArgs($args, __FUNCTION__);
+        $list = clone $this;
+        $list->rows = $this->getMatches($filters, true);
+        return $list;
+    }
+
+    public function exclude(...$args): static
+    {
+        $filters = $this->normaliseFilterArgs($args, __FUNCTION__);
+        $toRemove = $this->getMatches($filters);
+        $list = clone $this;
+        foreach ($toRemove as $id => $row) {
+            unset($list->rows[$id]);
+        }
+        return $list;
+    }
+
+    /**
+     * Return a copy of this list which does not contain any items with any of these params
+     */
+    public function excludeAny(...$args): static
+    {
+        $filters = $this->normaliseFilterArgs($args, __FUNCTION__);
+        $toRemove = $this->getMatches($filters, true);
+        $list = clone $this;
+        foreach ($toRemove as $id => $row) {
+            unset($list->rows[$id]);
+        }
+        return $list;
+    }
+
+    /**
+     * Return a new instance of the list with an added filter
+     */
+    public function addFilter(array $filterArray): static
```

Comment:
```suggestion
    public function addFilter($filterArray): static
```
Match DataList

Reply:
In that case I need to not strongly type the `$filters` argument in `getMatches` - which, I _suppose_ someone might be passing in some other type of iterable here.
Done.